### PR TITLE
[feature](Cloud) Support restore HDFS meta data in Meta Service && Add storage vault property

### PR DESCRIPTION
**Pull request by ByteYue**
## Proposed changes

Issue Number: close #xxx
This pr is part of `HDFS support for Compute Storage separation`. The main purpose of this is one that add support for hdfs meta data restore for Meta Service. And we introduce one item called `storage vault` which would be described in details in following pr.

TODO:
1. Reclaim the key value on FDB when the add hdfs info action failed.
2. Adjust BE code to refresh it's local obj/hdfs info using the new `get_obj_info` rpc interface
3. Adjust FE code to support Create table stmt with `storage vault` property.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: cloud/src/meta-service/keys.cpp [hunk 1]
```diff
@@ -115,11 +115,12 @@ static void encode_prefix(const T& t, std::string* key) {
...
         StatsTabletKeyInfo,
         JobTabletKeyInfo, JobRecycleKeyInfo, RLJobProgressKeyInfo,
-        CopyJobKeyInfo, CopyFileKeyInfo>);
+        CopyJobKeyInfo, CopyFileKeyInfo, StorageVaultKeyInfo>);
 
     key->push_back(CLOUD_USER_KEY_SPACE01);
     // Prefixes for key families
-    if        constexpr (std::is_same_v<T, InstanceKeyInfo>) {
+    if        constexpr (std::is_same_v<T, InstanceKeyInfo>
+                      || std::is_same_v<T, StorageVaultKeyInfo>) {
```
Comment (gavinchou) [inline cloud/src/meta-service/keys.cpp:123]: StorageVaultKeyInfo is not int the family of  InstanceKeyInfo.
it should be 
0x01 "storage_vault" $instance_id $resource_id

File: cloud/src/meta-service/meta_service_resource.cpp [hunk 2]
```diff
@@ -218,52 +271,158 @@ void MetaServiceImpl::get_obj_store_info(google::protobuf::RpcController* contro
...
-            obj_info.set_ak(std::move(plain_ak_sk_pair.first));
-            obj_info.set_sk(std::move(plain_ak_sk_pair.second));
+        if (auto ret = decrypt_and_update_ak_sk(obj_info, code, msg); ret != 0) {
+            return;
+        }
+    }
+    // Iterate all the resources to return to the rpc caller
+    for (const auto& resource_id : instance.resource_ids()) {
+        std::string storage_vault_k;
+        storage_vault_key({instance_id, resource_id}, &storage_vault_k);
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_resource.cpp:281]: use the form  `key = storage_vault_key()`;

File: cloud/src/meta-service/meta_service_resource.cpp [hunk 3]
```diff
@@ -218,52 +271,158 @@ void MetaServiceImpl::get_obj_store_info(google::protobuf::RpcController* contro
...
+        if (auto ret = fetch_all_storage_valut(std::move(storage_vault_k), instance_id, response,
+                                               txn_kv_.get(), code, msg);
+            ret != 0) {
+            return;
         }
     }
     response->mutable_obj_info()->CopyFrom(instance.obj_info());
 }
 
+static std::string next_avaiable_vault_id(const InstanceInfoPB& instance) {
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_resource.cpp:291]: lack of comment and UT, this function should be carefully tested.

**Comment by ByteYue** [comment]
run buildall

File: cloud/src/meta-service/meta_service_resource.cpp [hunk 4]
```diff
@@ -218,52 +271,158 @@ void MetaServiceImpl::get_obj_store_info(google::protobuf::RpcController* contro
...
+    } break;
+    case AlterObjStoreInfoRequest::ADD_HDFS_INFO: {
+        if (!request->has_hdfs()) {
+            code = MetaServiceCode::INVALID_ARGUMENT;
+            msg = "hdfs info is not found " + proto_to_json(*request);
+            return;
+        }
+    } break;
+    case AlterObjStoreInfoRequest::UNKNOWN:
+        break;
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_resource.cpp:425]: any error/warning msgs if we reach this branch?
it seems unexpected

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

File: cloud/src/meta-service/meta_service_resource.cpp [hunk 5]
```diff
@@ -218,52 +271,156 @@ void MetaServiceImpl::get_obj_store_info(google::protobuf::RpcController* contro
...
+            return;
         }
     }
     response->mutable_obj_info()->CopyFrom(instance.obj_info());
 }
 
+// The next avaiable vault id would be max(max(obj info id), max(vault id)) + 1.
+static std::string next_avaiable_vault_id(const InstanceInfoPB& instance) {
+    size_t vault_id = 0;
+    auto cmp = [](size_t prev, const auto& last) {
```
Comment (platoneko) [inline cloud/src/meta-service/meta_service_resource.cpp:293]:
```suggestion
    auto max = [](size_t prev, const auto& last) {
```

File: cloud/src/meta-service/meta_service_resource.cpp [hunk 6]
```diff
@@ -218,52 +271,156 @@ void MetaServiceImpl::get_obj_store_info(google::protobuf::RpcController* contro
...
+        return std::max(prev, last_id);
+    };
+    auto prev = std::accumulate(
+            instance.resource_ids().begin(), instance.resource_ids().end(),
+            std::accumulate(instance.obj_info().begin(), instance.obj_info().end(), vault_id, cmp),
+            cmp);
+    return std::to_string(prev + 1);
+}
+
+static int add_hdfs_storage_valut(InstanceInfoPB& instance, TxnKv* txn_kv,
```
Comment (platoneko) [inline cloud/src/meta-service/meta_service_resource.cpp:309]:
```suggestion
static int add_hdfs_storage_valut(InstanceInfoPB& instance, Transaction* txn,
```

**Review by platoneko — APPROVED** [review]

**Review by platoneko — CHANGES_REQUESTED** [review]
Plz resolve the comment

**Comment by ByteYue** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by platoneko — APPROVED** [review]

**Comment by ByteYue** [comment]
run buildall

**Pull request merged**